### DISCUSSION TITLE
v0.11.7 Testing checklist

### DISCUSSION THREAD

**Discussion by jorgeepc** [kubeshop/tracetest#2472] [Release testing]
# Testing Tracetest

This is the entire checklist on what we should do to assert that Tracetest is working fine on each version release. On each version release, we can copy the contents of this template checklist and open a Github Discussion to start each test.

On Tracetest, we work in two ways to test the entire system and guarantee that everything is working fine. One is through automatic tests, where we run unit, integration and end-to-end tests on Tracetest CLI, Web UI and Server.

Another source of tests is the manual tests that we execute on each release, simulating users, and checking if everything is ok on these tests.

## Manual Tests

On our manual tests, we aim to do some [sanity checks](https://en.wikipedia.org/wiki/Sanity_check) to see if the main features are working as expected. Usually, we run it on each Tracetest release.

### Tests to validate on Release Candidate

- Test server installation via CLI
  - [x] Docker Compose and no demo API
  - [x] Docker Compose and demo API
  - [x] Kubernetes and no demo API
  - [x] Kubernetes and demo API
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for docker install
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for k8s install

- Test DataStore provisioning and basic test
  - [x] Open WebUI and go to `/settings` page. The provisioned Data Store should be selected
  - [x] Create a test on WebUI that calls a demo API (like [Pokeshop](https://docs.tracetest.io/live-examples/pokeshop/overview) or [Open Telemetry Store](https://docs.tracetest.io/live-examples/opentelemetry-store/overview)). This test should fetch traces correctly and run without errors

- Test Tracetest examples (_**probably we could skip this step for the relase**_)
  - [x] [Jaeger example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-jaeger)
  - [x] [Tempo example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-tempo)
  - [x] [New Relic example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-new-relic)
  - [x] [Lightstep example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-lightstep)
  - [x] [OpenSearch example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-opensearch)
  - [x] [Elastic APM example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-elasticapm-with-elastic-agent)
  - [x] [SignalFX example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-signalfx)
  - [x] [Datadog example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-datadog)
  - [x] [Amazon X-Ray example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-amazon-x-ray)

- Test specific features added/changed on this release:
  - [x] Honeycomb integration. Test the following features: data store provisioning, webUI settings page, cli `get` | `apply` datastore command using honeycomb, run a test and validate data is being send to both Tracetest and Honeycomb
  - [x] Validate new CLI `environment` resource: `list environment`, `apply environment`, `get environment`, `delete environment`
  - [x] Validate CLI `test run` command using an environment `--environment`
  - [x] Validate webUI environment CRUD
  - [x] Validate a test run from the webUI using an environment
  - [x] Validate a transaction run from the webUI using an environment
  - [x] Validate test specs snippets from the webUI

### Tests to validate on Final Release

- Pipeline
  - [x] Check if our release pipeline on [Release Tracetest](https://github.com/kubeshop/tracetest/actions/workflows/release-version.yml) workflow on Github Actions worked correctly

- Test CLI updates
  - [x] MacOS via homebrew
  - [x] MacOS via curl script
  - [x] Windows via chocolatey
  - [x] Windows via manual download

- Docs
  - [x] Double check [Detailed installation](https://docs.tracetest.io/getting-started/detailed-installation) docs and see if everything is documented correctly

- Test specific features added/changed on this release:
  - [x] Add list of features here
  
## Automatic Tests

Today Tracetest has 3 main components: a WebUI, a CLI and a Server.

### Web UI

- **Unit tests**: Run by executing `npm test` on `./web` folder
- **End-to-end tests**: Run using [cypress](https://www.cypress.io/) against a temporary Tracetest created on Kubernetes. 

### CLI

- **Unit tests**: Run by executing `make test` on `./cli` folder

### Server

- **Unit tests**: Run by executing `make test` on `./server` folder
- **Integration tests**: Run along with some unit tests running `make test` on `./server` folder, but also done in a matrix test on Github actions, by executing the `test-examples` action.
- **End-to-end tests**: Run using Tracetest to test itself. We deploy two instances of Tracetest and use one to test API calls to another on Github actions, by executing the `trace-testing` action.

**Comment by xoscar** [comment]
Found two problems:
1. The tracetest-opensearch example is not able to fetch the trace (not sure if I did something wrong)
2. Using a grpc example when installing Tracetest using the CLI with the pokeshop demo doesn't work, All http examples work tho!

**Comment by xoscar** [comment]
Found another issue
- when clicking the test connection button when having honeycomb selected the app shows an errored notification

I'm fixing this

**Comment by xoscar** [comment]
Found another pre-existing bug:
- When running a transaction without an id form the CLI, it doesn't update the id field from the resource file, causing every consecutive run to create a new transaction entry
- Trying to delete a transaction with the above problem causes a foreign key issue in the DB
![Screenshot 2023-05-03 at 18 07 02](https://user-images.githubusercontent.com/11051031/236076724-38b97d60-c9ac-4751-a39f-6695d64bc397.png)

**Comment by xoscar** [comment]
Found a pre-existing bug with the environment form modal
- When trying to edit an environment by deleting one of the value entries, the update stays as disabled
